File: app/src/main/java/art/coded/wireframe/ui/list/ListViewModel.java
package art.coded.wireframe.ui.list;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import art.coded.wireframe.data.Element;
import art.coded.wireframe.data.ElementRepository;

import java.util.List;

public class ListViewModel extends ViewModel {

    private ElementRepository mRepository;
    private LiveData<List<Element>> mData;

    public void loadData(Application application) {
        mRepository = new ElementRepository(application);
        mData = mRepository.getAllElements();
    }

    public LiveData<List<Element>> getData() {
        return mData;
    }
    public void removeData(Element element) { mRepository.delete(element); }
}

File: app/src/ktx/java/art/coded/wireframe/viewmodel/ListViewModel.kt
package art.coded.wireframe.viewmodel

import androidx.lifecycle.*
import art.coded.wireframe.model.ElementRepository
import art.coded.wireframe.model.entity.Element
import kotlinx.coroutines.launch

private val LOG_TAG = ListViewModel::class.java.simpleName

class ListViewModel(private val repository: ElementRepository) : ViewModel() {

    val actionPending = MutableLiveData<Boolean>()

    fun getData(): LiveData<List<Element>> {
        var elements: LiveData<List<Element>>? = null
        actionPending.postValue(true)
        viewModelScope.launch {
            elements = repository.allElements
        }
        actionPending.postValue(false)
        return elements!!
    }

    fun removeData(element: Element) {
        actionPending.postValue(true)
        viewModelScope.launch {
            repository.delete(element)
        }
        actionPending.postValue(false)
    }
}

class ListViewModelFactory(private val repository: ElementRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return ListViewModel(repository) as T
    }
}